### PR TITLE
fix(Cognito): Update id token immediately after signIn

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:* https://github.com/aws-amplify/aws-sdk-ios/issues/2922

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift [hunk 1]
```diff
@@ -798,15 +804,32 @@ extension AWSMobileClient {
...
                         completionHandler(nil, error)
                         self.invokeSignInCallback(signResult: nil, error: error)
+                        self.tokenFetchLock.leave()
                     } else if let session = task.result {
+                        
                         completionHandler(self.userSessionToTokens(userSession: session), nil)
                         self.federationProvider = .userPools
                         if (self.currentUserState != .signedIn) {
-                            self.mobileClientStatusChanged(userState: .signedIn, additionalInfo: [:])
+                            // If the current state is not already signedIn, we might have reached here due to
```
[the remark this answers is not in the document]
Comment (royjit) [reply AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift:816]: Added comments.

File: AWSCore/Authentication/AWSCredentialsProvider.h [hunk 2]
```diff
@@ -258,6 +258,13 @@ Initializer for credentials provider with enhanced authentication flow. This is
  */
 - (AWSTask<NSString *> *)getIdentityId;
 
+/**
+ Update the identity id for an authenticated user. The api will try to update the existing identity id to match the current signed in logins map's identity id.
```
[the remark this answers is not in the document]
Comment (royjit) [reply AWSCore/Authentication/AWSCredentialsProvider.h:262]: Updated the documentation

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift [hunk 3]
```diff
@@ -830,6 +859,16 @@ extension AWSMobileClient {
         }
         return Tokens(idToken: idToken, accessToken: accessToken, refreshToken: refreshToken, expiration: userSession.expirationTime)
     }
+    
+    internal func updateIdentityId(_ logins: [String: String], completion: @escaping (Bool) -> Void) {
+        if let credentialsProvider = self.internalCredentialsProvider {
```
[the remark this answers is not in the document]
Comment (royjit) [reply AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift]: Changed

File: AWSCore/Authentication/AWSCredentialsProvider.m [hunk 4]
```diff
@@ -696,6 +696,98 @@ - (void)setUpWithRegionType:(AWSRegionType)regionType
     }];
 }
 
+#pragma mark - Helper method to update identity id immediately after signin
```
[the remark this answers is not in the document]
Comment (royjit) [reply AWSCore/Authentication/AWSCredentialsProvider.m]: Removed

File: AWSCore/Authentication/AWSCredentialsProvider.m [hunk 5]
```diff
@@ -696,6 +696,98 @@ - (void)setUpWithRegionType:(AWSRegionType)regionType
...
+        if (self.identityId == nil) {
+            return [[self getIdentityIdAPI:logins] continueWithSuccessBlock:^id (AWSTask *task) {
+                NSString *identityIdFromResponse = task.result;
+                if (identityIdFromResponse) {
+                    self.identityId = identityIdFromResponse;
+                    providerRef.identityId = identityIdFromResponse;
+                    return [AWSTask taskWithResult:[NSNumber numberWithBool:true]];
+                } else {
+                    return [AWSTask taskWithResult:[NSNumber numberWithBool:false]];
+                }
```
[the remark this answers is not in the document]
Comment (royjit) [reply AWSCore/Authentication/AWSCredentialsProvider.m]: Moved to separate function

**Review by palpatim — APPROVED** [review]

**Pull request merged**